### ISSUE TITLE
Including directories relative to source_root() escapes the build directory

### ISSUE THREAD

**Issue by codido** [mesonbuild/meson#1358]
If a meson.build script includes a subdir from a directory relative to source_root(), targets in that directory may be built in the source tree rather than the build directory.

For instance, say we have the following:
meson.build:
```
project('test', 'c')

subdir(join_paths(meson.source_root(), 'sub'))
```
sub/meson.build:
```
static_library('foo', 'foo.c')
```

Building the project will actually build libfoo inside the source tree:
```
$ meson build && cd build && ninja
The Meson build system
Version: 0.38.1
Source dir: /tmp/bug
Build dir: /tmp/bug/build
Build type: native build
Project name: test
Native c compiler: ccache cc (gcc 6.3.1)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Build targets in project: 1
[2/2] Static linking library /tmp/bug/sub/libfoo.a
```

**Comment by jpakkane** [comment]
> subdir(join_paths(meson.source_root(), 'sub'))

Please do not do this. Ever. It should not even work. Only use relative paths. This will be changed so that only relative paths work as soon as I have the time.

**Comment by nirbheek** [comment]
All file and directory paths if they are not absolute are assumed to be relative to the current directory, so you almost never need to use `meson.source_root()` or `meson.current_source_dir()`. *Especially* not in `subdir()`. I guess we don't check that the argument is not absolute, and that's a bug that will be fixed.

**Comment by codido** [comment]
I see - have I missed some warning in the documentation about this?

Any way around for including directories relative to the top directory instead of traversing with .., counting the number of directories that should be traversed back?
The use case for this is basically a shared subdir used by other subdirs with arbitrary depth. Something along the lines of root/sub1/sub2/sub3 root/sub1/sub2/sub4 including root/shared_sub.

**Comment by TingPing** [comment]
You only go into a subdir once, it creates whatever targets, then any subsequent files can reference those targets. It makes no sense to include a subdir multiple times.

**Comment by nirbheek** [comment]
In other words, all variables defined in the `meson.build` file in a directory entered with `subdir()` are available in all directories entered thereafter with `subdir()`. I'll update the reference manual to be clearer about this.

**Comment by codido** [comment]
@TingPing, @nirbheek the intention wasn't to include shared_sub more than once, but conditionally include it if need be. IOW, sub3 & sub4 above are mutually exclusive, depending on some configuration/option/etc.

Guess that scenario isn't supported, but can be worked around by including some of the decisions in the top level meson.build.

**Comment by nirbheek** [comment]
This is now fixed with #1376.
